### DISCUSSION TITLE
Mapping sub-lattice and their electrons

### DISCUSSION THREAD

**Discussion by ArunPsiog** [AgnostiqHQ/covalent#1096] [Q&A] [answered]
### Description
Current database, it doesn’t have a direct reference to sub-lattice (in lattices table) and their electrons (in electrons table).

### Observation
The _parent lattice ID_ of an electron or sub-lattice refers to the lattice in which each of those is invoked, but the database doesn't have any direct references to **sub-lattice** and their **electrons**. Hence, we are facing issues while referring sub lattice to their electron for Graph API.

### Suggestion
We can introduce new column in electrons, which will refer to the immediate sub-lattice or lattice (in the absents of sub-lattice).

### Change Implementation

**Database changes,**
    1. covalent/_data_store/models.py, To modify for adding the new column
    2. covalent_ui/api/v1/database/schema/electrons.py, To modify for adding the new column

**Data layer changes,**
    1. insert_electrons_data in write_result_to_db.py to add new column data.
    2. Electrons in electron_dal.py. To add dispatch id in electron id.

### Queries
1. Can a sub-lattice be a child node of another sub lattice?
2. How to refer a sub-lattice which is child node an another sub-lattice?

cc: @mshkanth, @Prasy12

**Comment by cjao** [comment]
Hi @ArunPsiog, with the current schema, one can easily find a sublattice's parent lattice via the `Lattices.electron_id` column, ~~but I agree that it is inconvenient to go the other way. If I understand correctly, you are proposing a `sublattice_dispatch_id` column in the electrons table which would be blank for ordinary electrons and populated with the sublattice dispatch id for sublattices, thereby making the relationship between a lattice and its sublattices bi-directional. I think such a change would be quite natural; it would greatly simplify querying all the sublattices of a lattice.~~

**Edit**: based on today's meeting I see that I misunderstood the initial question

**Comment by cjao** [comment]
Regarding your questions:
1) Sublattices can be nested, but every sublattice belongs to exactly one top-level lattice. The set of lattices forms a forest whose roots are the top-level lattices.
2) Once you have a procedure to determine the dispatch ids of the first level of sublattices within a transport graph, it can be applied iteratively to nested sublattices.

**Reply by ArunPsiog** [reply]
Hi @cjao , Thanks for your reply 
 
We could have a new column named _immediate_parent_lattice_id_, which will refer to the immediate parent of the electron/sub-lattice.
- For an electron which is a direct child of a lattice, the new field can record immediate lattice id or dispatch id(it can be the same as parent lattice id). 
- For a sub-lattice(sl2), which is a direct child of another sub-lattice(sl1). The new field can record immediate sub-lattice id or dispatch id(i.e sl1). 


cc: @mshkanth, @Prasy12

**Comment by cjao** [comment]
From today's meeting, the main problem is that the `parent_lattice_id` of a sublattice electron always points to the root lattice, not to the immediate parent as intended. @Prasy12 would the new column still be needed after this implementation error is corrected?

**Reply by ArunPsiog** [reply]
Hi @cjao, 

We might still need some reference to the root lattice in electron table, otherwise we need to iterate for every sub-lattice and then show the complete graph to the user, which is an inefficient way.

Lets assume we have multi level sub-lattice, as below image

![Screenshot from 2022-08-24 20-34-00](https://user-images.githubusercontent.com/106462226/186453882-e4691ee6-3733-4f2c-9ef1-545f4a97c336.png)

L - Root Lattice
SL1 - Sub-lattice
SL2 - Sub-lattice (child of SL1)
d1,d2,d3 are the reference to immediate sub-lattice

Case 1: If we have Immediate parent lattice id,
- At first we will get immediate child of _root lattice_, then we need to identify electron's type, if its a _sub-lattice_, again we need to get its children.
- For this case we need to iterate twice as we have two sub- lattice.

Case 2: If we have immediate parent and root reference,
We can get all electrons from electron table with the help of root reference. 

 CC: @mshkanth, @Prasy12

**Reply by cjao** [reply]
Hi @ArunPsiog, I was under the impression that the user would expand one level of sublattices at a time:
* (L, e, SL1)  (initial view)
* (L, e, (e, SL2))  (after clicking on SL1)
* (L, e, (e, (e, e)))  (after clicking on SL2)

If this is the case, wouldn't the iterative approach suffice? Each sublattice electron could be expanded with two SELECTs (to get the nodes for the sublattice graph) and a JOIN (for the edges).

**Reply by ArunPsiog** [reply]
Hi @cjao,

My assumption is that once a user navigates to the graph page, the user can view only the first level children of the root lattice, but not the entire graph (without the children of the sub-lattice at that level). If a user needs to view the children nodes of a sub-lattice, the user can expand the sub-lattice, which will internally trigger an API call to get children of that sub-lattice.

CC: @Prasy12 , @mshkanth

**Comment by cjao** [comment]
I tried reproducing the issue with this script:
```python
import covalent as ct

@ct.electron
def sub_sub_task(x):
    return x**2

@ct.electron
@ct.lattice
def sub_sub_workflow(x):
    # 1 -> 0
    return sub_sub_task(x)

@ct.electron
@ct.lattice
def sub_workflow(x):
    # 1 -> 0
    return sub_sub_workflow(x)

@ct.lattice
def workflow(x):
    # 1 -> 0
    return sub_workflow(x)

dispatch_id = ct.dispatch(workflow)(3)
print(dispatch_id)
print(ct.get_result(dispatch_id, wait=True))
```
Starting with an empty database, I get the following table entries:

Lattices (`id|dispatch_id|electron_id|completed_electron_num|electron_num`):
```
>>> sqlite3 ~/.local/share/covalent/dispatcher_db.sqlite "select id, dispatch_id, electron_id, completed_electron_num, electron_num from lattices"

1|f347eef4-650e-43e9-b938-3e342380ede6| |2|2
2|68dba942-48f0-401e-91a0-4e0bc1a3a034|1|2|2
3|93cf01bf-b12d-4ce2-b19f-576d41811415|3|2|2
```

Electrons (`id|parent_lattice_id|type`)
```
>>> sqlite3 ~/.local/share/covalent/dispatcher_db.sqlite "select id, parent_lattice_id, transport_graph_node_id, type from electrons"

1|1|0|sublattice
2|1|1|parameter
3|2|0|sublattice
4|2|1|parameter
5|3|0|function
6|3|1|parameter
```

```
>>> sqlite3 ~/.local/share/covalent/dispatcher_db.sqlite "select electron_id, parent_electron_id from electron_dependency"
1|2
3|4
5|6
```

The `parent_lattice_id` field is being populated as designed:

Starting with the top-level lattice (`1`), I see that electron `1` is a sublattice. If I want to expand the graph of that sublattice, I query the nodes with  `parent_lattice_id=2` and their corresponding edges. To expand the electron `3` representing a sub-sublattice, I query the nodes with `parent_lattice_id=3`.

@ArunPsiog clearly I'm missing something here.  Is this example too simple?

**Reply by ArunPsiog** [reply]
@cjao, I ran your script, It worked fine. I think there are some issues with my script.

**Comment by mshkanth** [comment]
Hi @santoshkumarradha - Got a download of yesterday's between all of you. Want to quickly clarify expectation reg sub-lattices below - 

**1** 
Hope the expectation is that when a user loads the graph, all nodes will be collapsed first. If need be, the user will expand/collapse each sub-lattice and the sub-lattice's electrons will be loaded on-demand. ie Nowhere in the system are we expected a direct full load of the graph (including all sub-lattices & nested sub-lattices). 

**2**

![image](https://user-images.githubusercontent.com/23229840/186662711-0b83f998-a058-41ed-915c-b82803705829.png)

In the new sub-lattices tab where the list of sub-lattices are expected to be shown, again is the expectation only to show root level sub-lattices or are all nested sub-lattices expected to be shown? If it is the former, then there can be some drill down option for a user to drill into one sub-lattive and see the nested sub-lattices under it. If it is the later, then pls see the explanation below - 

In the current database, the `electrons.parent_lattice_id` stores the immediate parent. Hence, it is easy to pick 1 sub-lattice and see all it's 1st level children. However take a nested sub-lattice scenario (Sub-lattice inside Sub-lattice) - To flatten out the entire hierarchy and also show all nested child sub-lattices, the query needs to loop between the `lattices` & `electrons` back & forth to retrive the whole hierarchy. To aid this process, we request that a `root_lattice_id` column also be maintained in the `electrons` table that will avoid the back & forth loop. This request however depends very much on the requirement 1 and 2 that I have pasted above. Pls let us know. 

Cc @ArunPsiog @Prasy12

**Reply by santoshkumarradha** [reply]
Hi @mshkanth 

For 1. the ask is indeed to have atleast an option to open up all sublattices (but in lieu of feature addition this can be a v2 work) provided they are generated. (Note that sublattices are dynamic) 

For 2. We need all the sublattice shown in list (it does not matter if it's nested for the left side bar, the concept of hierarchy does not need to exist for left bar). User needs to have the status of all "dispatched" sublattices and a way to get to it in the graph by clicking the left bar. 

So with the requirement I assume there is a need for that column ? I will let @wjcunningham7 / @cjao  and @Prasy12 have the discussion about the specifics of the db modification.

**Reply by cjao** [reply]
>For 2. We need all the sublattice shown in list (it does not matter if it's nested for the left side bar, the concept of hierarchy does not need to exist for left bar). User needs to have the status of all "dispatched" sublattices and a way to get to it in the graph by clicking the left bar.

If there is indeed a requirement to list all the sublattices under a root lattice, then a `root_lattice_id` column would indeed make that query more efficient. Will the UI allow treating sublattices also as root lattices for their own nested sublattices? In other words, consider a hierarchy like
```
       L
    /   \
  SL1   e
  /   \ 
SL2   SL3
```
If I understand correctly, the workflow view for L would list SL1, SL2, and SL3 as sublattices. My question is if SL1 would also appear in the main dispatch list as it did previously. If so, when the user clicks on SL1, then the workflow view for SL1 would list SL2 and SL3 as sublattices. With this UX, the `root_lattice` of a given sublattice would depend on which sublattice the user considers "top level".

**Reply by Prasy12** [reply]
> > For 2. We need all the sublattice shown in list (it does not matter if it's nested for the left side bar, the concept of hierarchy does not need to exist for left bar). User needs to have the status of all "dispatched" sublattices and a way to get to it in the graph by clicking the left bar.
> 
> If there is indeed a requirement to list all the sublattices under a root lattice, then a `root_lattice_id` column would indeed make that query more efficient. Will the UI allow treating sublattices also as root lattices for their own nested sublattices? In other words, consider a hierarchy like
> 
> ```
>        L
>     /   \
>   SL1   e
>   /   \ 
> SL2   SL3
> ```
> 
> If I understand correctly, the workflow view for L would list SL1, SL2, and SL3 as sublattices. My question is if SL1 would also appear in the main dispatch list as it did previously. If so, when the user clicks on SL1, then the workflow view for SL1 would list SL2 and SL3 as sublattices. With this UX, the `root_lattice` of a given sublattice would depend on which sublattice the user considers "top level".

The SL1 would not appear inside the main dispatch list after this UI change. Only L will appear on the main dispatch list and SL1,SL2 and SL3 needs to be shown inside the left sidebar view as per the screenshot @mshkanth has posted.

**Reply by Prasy12** [reply]
> For 2. We need all the sublattice shown in list (it does not matter if it's nested for the left side bar, the concept of hierarchy does not need to exist for left bar). User needs to have the status of all "dispatched" sublattices and a way to get to it in the graph by clicking the left bar.
> 

@cjao When you look at 2, there is also a need for the sublattice part of the graph to be highlighted when clicked upon, in this case we might actually need to end up showing the whole graph on inital load with all sublattices expanded. Only if below is shown we can zoom in on SL3 when clicked from the left sidebar,
L
/
SL1 e
/ \
SL2 SL3

**Reply by cjao** [reply]
@Prasy12 thanks for clarifying. Since sublattices are not going to be treated as full dispatches as before, I think a `root_lattice_id` column makes a lot of sense.

**Reply by Prasy12** [reply]
Thanks @cjao Can you let us know when the column addition  can be done? It'll be easier for us as well to plan and estimate the whole feature end-end.
If we are in agreement that a new column will be add, for our development purpose we can create our own mock and work on that until you guys finish development and push the feature. Let me know if you have any different thoughts.

cc : @wjcunningham7 @santoshkumarradha @ArunPsiog

**Reply by wjcunningham7** [reply]
I agree this extra field makes sense. @cjao can you work with @AlejandroEsquivel to scope this and determine the timeline?

**Reply by cjao** [reply]
Working on this now. I have related suggestion -- since `root_lattice_id` is really a lattice-level property, would it fit better in the Lattices table?

**Reply by cjao** [reply]
@ArunPsiog if Lattices contains a `root_lattice_id` column, it would also be easy to discover all the sublattices under a top-level lattice.

**Reply by ArunPsiog** [reply]
@cjao, 
Adding `root_lattice_id` in Lattices, would simplify the query to get _**sub-lattice**_, If electrons contain `root_lattice_id` that would help us to get **_sub-lattices_** (from `electrons.type`) and _**electrons**_  associated with a particular workflow. 

I would still suggest implementing  `root_lattice_id` in Electrons, where it's simpler to query **_sub-lattices_** and **_electrons._**. Please let us know if you have another opinion? Thanks. 

CC: @Prasy12

**Comment by ArunPsiog** [comment]
Hi @cjao,

Links between,
- Lattice (L) -> Sub-lattice (SL)
- Sub-lattice (SL) -> Sub-sub-lattice (SSL) 

Are missing in electron_dependency table.  I have used this below script to create sub-lattices workflow,
```
import covalent as ct

@ct.electron
def sub_sub_task(x):
    return x**2

@ct.electron
@ct.lattice
def sub_sub_workflow(x):
    # 1 -> 0
    return sub_sub_task(x)

@ct.electron
@ct.lattice
def sub_workflow(x):
    # 1 -> 0
    return sub_sub_workflow(x)

@ct.lattice
def workflow(x):
    # 1 -> 0
    return sub_workflow(x)

dispatch_id = ct.dispatch(workflow)(3)
print(dispatch_id)
print(ct.get_result(dispatch_id, wait=True))
 ```
With a empty database. The workflow creates, 
- 3 lattices in lattices table
- 6 electrons in **electrons** table 
- 3 links in **electron_dependency** table. 

Actual,
```
           L
             \
      SL      P
        \ 
  SSL    P
  /
P

```

Expected,
```
           L
         /   \
      SL      P
      /  \ 
  SSL    P
  /
P

```


/, \ - Represent links present in electron_dependency table
L - lattice
SL - sub-lattice of lattice (L)
SSL - sub-lattice of sub-lattice (SL) 

Refer below screenshot for database entries, 

**Lattices table:** 
![image](https://user-images.githubusercontent.com/106462226/186897110-61f75420-1ebf-403c-9ee7-960d6bac690d.png)

**Electrons table:**
![image](https://user-images.githubusercontent.com/106462226/186897340-00b827c9-f9dc-4197-a44f-c88b6099fdae.png)

**Electron Dependency table**
![image](https://user-images.githubusercontent.com/106462226/186897657-40adf4c8-7b14-4745-b730-ab0fc5465db0.png)

While adding the new field (root_lattice_id), Can you add this missing links in electron dependency table?

CC: @Prasy12 , @mshkanth

**Reply by cjao** [reply]
Hi @ArunPsiog,

Lattices and sublattices are represented by graphs, not nodes. In this example,  the overall workflow comprises of three graphs:

L:
```
        2 [parameter]
       /
[1 (SL1)]
       
```
SL1:
```
        4 [parameter]
       /
[3 (SL2)]
       
```
SL2:
```
        6 [parameter]
       /
[5 (sub_sub_task)]
       
```

Note that parameter node 2 is an input to the sublattice SL1 and not directly connected to any node in SL1. 

In fact, the same input parameter to a lattice may be stored in multiple parameter nodes. Consider the following example
```python
@ct.electron
def task(x):
    return x**2

@ct.lattice
def workflow(x):
    res1 = task(x)
    res2 = task(x)
    res3 = task(x)
    ...
```
Even though all tasks in the workflow are fed the same lattice input parameter, the transport graph will contain *three* parameter nodes with value x.

**Reply by ArunPsiog** [reply]
Hi @cjao ,

I ran script from [comment](https://github.com/AgnostiqHQ/covalent/discussions/1096#discussioncomment-3483012). On the graph page,

![image](https://user-images.githubusercontent.com/106462226/187145956-93b6f253-972e-4285-a7f7-1b209fe17a27.png)

Here, the user are able to see 3 different nodes, and their links to parameters. Is this the expected behavior?. When a user navigates into graph page by clicking on dispatch and all **electrons** and **sub-lattices** (including sub-lattices from different levels) were populated. 

Assuming **_root_lattice_id_** is implemented in **_electrons_**, when a user select a **_dispatch_id_**, 
* `lattices.id` of that dispatch_id is fetched from lattice
*  With that `lattices.id`, electrons are selected (`lattices.id == electrons.root_lattice_id`).
* The selected electrons, electron links were fetched from electron_dependency ( `electrons.id == electrons.parent_electron_id`).

**Observation**
In the current implementation, the links between two electrons were maintained at electron_dependency. For multi level sub-lattices, links such as `1(sub_workflow) -> 3(sub_sub_workflow)` and `3(sub_sub_workflow) -> 5(sub_sub_task)`, were missing in electron_dependency. Hence we are getting 3 different graphs.

Graphs,
```

        6 [parameter]
       /
[5 (sub_sub_task)]


        4 [parameter]
       /
[3 (sub_sub_workflow)]

        2 [parameter]
       /
[1 (sub_workflow)]
```
For single graph representation,
```
        6 [parameter]
       /
[5 (sub_sub_task)]
      \   
  m2   \         4 [parameter]
        \       /
        [3 (sub_sub_workflow)]
             \
     m1       \       2 [parameter]
               \     /
              [1 (sub_workflow)]

```
The m1, m2 are the missing link in electron_dependency.

Query
Is this 3 graph approach fine? Where users view three different graphs for a single workflow.
or
Is it better to have a single graph approach? If so, can you suggest any alternative ways to identify,  `3(sub_sub_workflow)` parent electron of `1(sub_workflow)` and  `5(sub_sub_task)`parent of  `3(sub_sub_workflow)`?

CC: @Prasy12

**Reply by Prasy12** [reply]
Without the links that @ArunPsiog has mentioned, the concept of expanding/collapsing of a sublattice might not be possible as far as we have researched. Please do let us know if you think different and if there is a mechanism to do it.

**Reply by cjao** [reply]
> Hi @cjao ,
> 
> I ran script from [comment](https://github.com/AgnostiqHQ/covalent/discussions/1096#discussioncomment-3483012). On the graph page,
> 
> ![image](https://user-images.githubusercontent.com/106462226/187145956-93b6f253-972e-4285-a7f7-1b209fe17a27.png)
> 
> Here, the user are able to see 3 different nodes, and their links to parameters. Is this the expected behavior?. When a user navigates into graph page by clicking on dispatch and all **electrons** and **sub-lattices** (including sub-lattices from different levels) were populated.

Hi @ArunPsiog, perhaps we are running different branches. This is what I see for that workflow:
![Screenshot from 2022-08-29 08-37-02](https://user-images.githubusercontent.com/15801897/187202711-d1b03584-3aaf-41e3-8257-3adbc30b3879.png)

**Reply by Prasy12** [reply]
Yes @cjao the graph that @ArunPsiog  has posted is once we add the root_lattice_id in the electron table and query the graph for the root 
 dispatch.

**Reply by cjao** [reply]
> **Observation** In the current implementation, the links between two electrons were maintained at electron_dependency. For multi level sub-lattices, links such as `1(sub_workflow) -> 3(sub_sub_workflow)` and `3(sub_sub_workflow) -> 5(sub_sub_task)`, were missing in electron_dependency. Hence we are getting 3 different graphs.
> 
> Graphs,
> 
> ```
> 
>         6 [parameter]
>        /
> [5 (sub_sub_task)]
> 
> 
>         4 [parameter]
>        /
> [3 (sub_sub_workflow)]
> 
>         2 [parameter]
>        /
> [1 (sub_workflow)]
> ```
> 
> For single graph representation,
> 
> ```
>         6 [parameter]
>        /
> [5 (sub_sub_task)]
>       \   
>   m2   \         4 [parameter]
>         \       /
>         [3 (sub_sub_workflow)]
>              \
>      m1       \       2 [parameter]
>                \     /
>               [1 (sub_workflow)]
> ```
> 
> The m1, m2 are the missing link in electron_dependency.

Hi @ArunPsiog, the transport graph represents the flow of data in a workflow, and edges have a precise meaning: two nodes are linked if the output of one node is an input of another node.  In this example, an edge from node 1 to node 3 would be misleading since `sub_sub_workflow` does not depend on the output of `sub_workflow`.

**Reply by ArunPsiog** [reply]
@cjao Is there any other way to get the link between electron 1 and 3?

**Reply by cjao** [reply]
@ArunPsiog Electrons 1 and 3 are not directly related given the current transport graph semantics.

**Reply by cjao** [reply]
>Without the links that @ArunPsiog has mentioned, the concept of expanding/collapsing of a sublattice might not be possible as far as we have researched. Please do let us know if you think different and if there is a mechanism to do it.

Hi @Prasy12, in this example, would it be possible to enlarge electron 1 and render the graph of SL1 inside the enlarged electron? Something like
![tg](https://user-images.githubusercontent.com/15801897/187209938-230bea34-c82b-4dc8-9144-63b6566c15e5.png)

**Reply by cjao** [reply]
@santoshkumarradha is this the UI you had in mind?

**Reply by Prasy12** [reply]
We have a left side bar which has list of sublattices, one way to counter this would be to on the inital load we have 1 and 2. When a user click on the sublattices, we just show 3 and 4.

**Reply by Prasy12** [reply]
There is also one small problem with on demand expansion/collapse. The idea of the expansion and collapse is when a user clicks on a sublattices on the left side bar the expansion happens. Imagine if there are multi level sublattices.
 L-SL1-SL2-SL3... In case if a user clicks SL3 on the left hand bar, we might have to first expand SL1 and then SL2 and then SL3 and put them all in a rounded box which might not be straighforward at all. @santoshkumarradha @cjao

**Reply by Prasy12** [reply]
> > Without the links that @ArunPsiog has mentioned, the concept of expanding/collapsing of a sublattice might not be possible as far as we have researched. Please do let us know if you think different and if there is a mechanism to do it.
> 
> Hi @Prasy12, in this example, would it be possible to enlarge electron 1 and render the graph of SL1 inside the enlarged electron? Something like ![tg](https://user-images.githubusercontent.com/15801897/187209938-230bea34-c82b-4dc8-9144-63b6566c15e5.png)

I will give this a bit of a thought and get back to you @cjao

**Reply by cjao** [reply]
>If there is no relationship between 1 and 3 we would not be able to get this box and include 1,3 and 4 together. There has to be someway to know these are related to get that bounding box and resize electron 1.

**Caution**: The following is all naive thinking and assumes you have a great deal  of control over the graph layout with your UI framework.

The size of electron 1 would depend only on the size of the bounding box for (3, 4) and not its origin. Could one first resize electron 1 and then adjust the origin of the bounding box? 

In general, suppose one has a nested hierarchy L-SL1-SL2-...-SLn and the user clicks on SLn in the side bar. Would a two-pass approach like the following be feasible?
1) Determine the sizes of the bounding boxes for each sublattice in a bottom-up fashion: For each k starting at n: from the size of the bounding box for SL(k), compute the sizes of the expanded electron in SL(k-1) and the overall bounding box for SL(k-1).
2) Now that one knows the sizes of all the expanded electrons in the hierachy, composite the graphs top-down: for each k starting at 1, render the graph of SL(k), and align the origin of SL(k+1) with the expanded electron for SL(k).

**Reply by cjao** [reply]
@santoshkumarradha I'm out of my depth here. Please help when you get a chance.

**Reply by Prasy12** [reply]
@cjao Appreciate your valuable inputs and thoughts. Will go back and see if there is an easier way to do this as well. 
@santoshkumarradha  If you can confirm if this is the expectation, it would be great.

![image](https://user-images.githubusercontent.com/54540812/187231650-f91dcdd6-35f3-453f-89d2-fe6a11769ef4.png)

**Comment by Prasy12** [comment, marked as answer]
@cjao There is one expectation to have this below format where the whole graph along with the sublattices need to be shown. 
![image](https://user-images.githubusercontent.com/54540812/186925256-35f8b68a-237a-4733-a270-659b3c2929d4.png)

With the current schema, do you think this might be a possible option to implement considering we don't have the links in the electron dependecny table as @ArunPsiog had pointed out
? 

If not, @santoshkumarradha we might need to think of an alternative to show the sublattices in the graph view.

**Reply by cjao** [reply]
@Prasy12 It's possible if you are willing to enclose the expanded sublattice in a box or some other visual aid to indicate that the edge from L connects to the original sublattice *electron* and not to any node in the expanded sublattice graph.
```
         L
      /     \
 -----        P
| SL |
| __ |

**Reply by ArunPsiog** [reply]
@cjao , If we consider the above workflow in the [comment](https://github.com/AgnostiqHQ/covalent/discussions/1096#discussioncomment-3483012),

```
              [1 (sub_workflow)]
                 /          \
                /              2 [parameter]
           [3 (sub_sub_workflow)]
```
**_sub_workflow_**, is the sub-lattice of **_workflow_** and **_sub_sub_workflow_** is the sub-lattice of **_sub_workflow_**. As per our understanding, with the current schema, there are no direct references stating the relationship between `electrons.id` 1 and  `electrons.id` 3.  Without that reference, graphs look like,

```
              [1 (sub_workflow)]
                          \
                             2 [parameter]
           [3 (sub_sub_workflow)]
```

And if the user tries to expand the graph,

```
              [1 (sub_workflow)]
                          \
                             2 [parameter]
           [3 (sub_sub_workflow)]
               \
                4 [parameter]
```
Is this the expected behavior?
 
CC: @Prasy12

**Reply by cjao** [reply]
@ArunPsiog as I commented above, it would be misleading to draw an edge from 1 -> 3 because there is no input-output relationship between the those two nodes. A more accurate representation would look something like
```
[2 (parameter)]
       \
         \
-------------------------------
|      [4 (parameter)]          |
|               \               |
|       [3 (sub_sub_workflow)]  |      [1 (sub_workflow)]
|                               |
| ---------------------------   |
```

**Reply by ArunPsiog** [reply]
@cjao , In your example,

```
[2 (parameter)]
       \ 
         \
-------------------------------
|      [4 (parameter)]          |
|               \               |
|       [3 (sub_sub_workflow)]  |      [1 (sub_workflow)]
|                               |
| ---------------------------   |

```

How do you link electron 2 (Which is a parameter) -> electron 4 and 3( electrons from sub-lattice)? Is there any database link which we missed?

**Reply by cjao** [reply]
@ArunPsiog In this example, electron 2 is linked to electron 1, which is now represented by the enlarged box. The graph of the sublattice is intended to be superimposed over electron 1's box.

**Reply by cjao** [reply]
Is this possible without a major redesign of the front-end code? I would imagine you first render the sublattice graph and then compute the bounding box of the sublattice graph in order to resize electron 1 appropriately.

**Comment by santoshkumarradha** [comment]
Yes indeed that’s the ask, the edge will be connected to a graph (which
when I expanded will be an electron of the sublattice.

On Fri, Aug 26, 2022 at 7:59 PM Casey Jao ***@***.***> wrote:

> @Prasy12 <https://github.com/Prasy12> It's possible if you are willing to
> enclose the expanded sublattice in a box or some other visual aid to
> indicate that the edge from L connects to the original sublattice
> *electron* and not to any node in the expanded sublattice graph.
>
>          L
>       /     \
>  -----        P
> | SL |
> | __ |
>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AgnostiqHQ/covalent/discussions/1096#discussioncomment-3484336>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG74SGBYE4VLLZMFPPNNQEDV3DIGHANCNFSM57APJ4YQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 

Santosh Kumar Radha

Agnostiq Inc.
